### PR TITLE
[DISCUSSION] Consume Folly as submodule

### PR DESCRIPTION
**Pull request by JunielKatarn**
Open for discussion before code review.

Addressing the offline dependency discussion over Folly, this is part 1 of a proposed change:
1. For RNW developers, consume Folly as a Git submodule
2. For RNW development users, consume Folly as part of the RNW npm, copied from this repo's submodule directory.

- Adds Folly as a Git submodule.
- Removes redundant download logic in MSBuild targets.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/6828)

### CODE REVIEW

File: vnext/Folly/Folly.vcxproj [hunk 1]
```diff
@@ -267,23 +267,9 @@
...
   </Target>
   <ItemGroup>
     <TemporaryFollyPatchFiles Include="$(MSBuildThisFileDirectory)\TEMP_UntilFollyUpdate\**\*.*" />
   </ItemGroup>
   <Target Name="Deploy" />
-
-  <!-- Reenable this task if we need to temporarily replace any folly files for fixes, while we wait for PRs to land in folly -->
-  <Target Name="ApplyFollyTemporaryPatch" BeforeTargets="PrepareForBuild" DependsOnTargets="UnzipFolly">
-    <Copy DestinationFiles="@(TemporaryFollyPatchFiles->'$(FollyDir)folly\%(RecursiveDir)%(Filename)%(Extension)')" SourceFiles="@(TemporaryFollyPatchFiles)" />
-  </Target>
```
Comment (asklar) [inline vnext/Folly/Folly.vcxproj:287]: we have forked files, these still need to override the version from folly

File: .gitmodules [hunk 2]
```diff
@@ -0,0 +1,5 @@
+[submodule "Folly"]
+	path = GitModules/Folly
+	url = https://github.com/facebook/folly.git
+	branch = master
```
Comment (asklar) [inline .gitmodules:4]: this needs to be a LKG version, not master

**Review by asklar — COMMENTED** [review]
creating a comment as a reminder that this needs package.json to include the folly files

**Comment by asklar** [comment]
need to update rnw-dependencies to ensure that git clone happens with submodules

**Review by asklar — CHANGES_REQUESTED** [review]
.

File: .gitmodules (hunk 2 above)
Comment (JunielKatarn) [reply .gitmodules:4]: The submodules config does not allow specifying a tag or commit.
LKGs are derived from the master branch (if this has changed, this file can be adjusted accordingly).

The actual version "selection" is done by going to the submodule directory and fetching/checking out the desired commit.

Currently, the checked-out version is `v2020.10.05.00`.

File: vnext/Folly/Folly.vcxproj (hunk 1 above)
Comment (JunielKatarn) [reply vnext/Folly/Folly.vcxproj:287]: I was unaware of such fork.
Will revert/adapt accordingly.

**Comment by dannyvv** [comment]
Since we have husky for git hooks, can we make that handle the submodule operations for us? i.e. on all operations where the working tree changes that it auto updates the working tree of the submodule?

**Comment by JunielKatarn** [comment]
> Since we have husky for git hooks, can we make that handle the submodule operations for us? i.e. on all operations where the working tree changes that it auto updates the working tree of the submodule?
> I generally always have trouble with git hooks running and slowing things down, but I have also many times spent hours investigating issue when I forgot to sync the submodules after switching branches.

I'm not proficient with husky, but if it's an incremental change, we can add it.
Personally, I'm very submodule aware (and PowerShell's `posh-git` module helps), but I'm on board taking the least intrusive approach for everyone.

**Comment by NickGerleman** [comment]
Before we dig too far into implementation, it seems worth figuring out if we want to do this vs source copies vs other alternatives. There's been some parallel discussion in the email thread over approach.

**Comment by JunielKatarn** [comment]
> Before we dig too far into implementation, it seems worth figuring out if we want to do this vs source copies vs other alternatives. There's been some parallel discussion in the email thread over approach.

For context, this branch was already up before the discussion started.
Note, I added a `[DISCUSSION]` prefix to this PR's title.
The purpose is to have a proof of concept ready, so we can eventually merge it, update it based on the discussion's outcome, or just discard it.

**Comment by JunielKatarn** [comment]
> need to update rnw-dependencies to ensure that git clone happens with submodules

Done.

**Pull request closed without merging**